### DISCUSSION TITLE
How do the duals get initialized?

### DISCUSSION THREAD

**Discussion by mahonbt1** [coin-or/Ipopt#685] [Q&A] [answered]
We are seeing minor differences in iterations between different builds of our solver that should be exactly the same.  The best we can tell is that the Dual variables are going uninitialized, or something to that effect.

I wanted to ask how `init_lambda` and `lambda` work inside of the TNLP's get_starting_point() method.  Is there an option we should be setting at startup to get this under control?

**Comment by svigerske** [comment, marked as answer]
By default, Ipopt will calculate some initial value for the dual values. It uses a least-square estimate for constraint multipliers and a constant for bound multipliers, see the various [Initialization options](https://coin-or.github.io/Ipopt/OPTIONS.html#OPT_Initialization).

If you enable [`warm_start_init_point`](https://coin-or.github.io/Ipopt/OPTIONS.html#OPT_warm_start_init_point), then Ipopt will call `get_starting_point()` with `init_lambda = true` and you are supposed to fill `lambda` with a suitable value.
